flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategy.java

```diff
@@ -0,0 +1,296 @@
+		DeploymentOption updateOption = new DeploymentOption(true);
+		DeploymentOption nonUpdateOption = new DeploymentOption(false);
+
+		for (SchedulingExecutionVertex schedulingVertex : schedulingTopology.getVertices()) {
+			DeploymentOption option = nonUpdateOption;
+			for (SchedulingResultPartition srp : schedulingVertex.getProducedResultPartitions()) {
+				SchedulingIntermediateDataSet sid = intermediateDataSets.computeIfAbsent(srp.getResultId(),
+					(key) -> new SchedulingIntermediateDataSet());
+				sid.addSchedulingResultPartition(srp);
+				if (srp.getPartitionType().isPipelined()) {
+					option = updateOption;
+				}
+			}
+			deploymentOptions.put(schedulingVertex.getId(), option);
+
+			if (schedulingVertex.getConsumedResultPartitions().isEmpty()) {
+				// schedule vertices without consumed result partition
+				executionVertexDeploymentOptions.add(
+					new ExecutionVertexDeploymentOption(schedulingVertex.getId(), option));
+			}
+		}
+
+		schedulerOperations.allocateSlotsAndDeploy(executionVertexDeploymentOptions);
+	}
+
+	@Override
+	public void restartTasks(Set<ExecutionVertexID> verticesToRestart) {
+		// increase counter of the dataset first
+		for (ExecutionVertexID executionVertexId : verticesToRestart) {
+			final SchedulingExecutionVertex schedulingVertex = schedulingTopology.getVertex(executionVertexId)
+				.orElseThrow(() -> new IllegalStateException("can not find scheduling vertex for " + executionVertexId));
+
+			for (SchedulingResultPartition srp : schedulingVertex.getProducedResultPartitions()) {
+				if (srp.getPartitionType().isBlocking() && DONE.equals(srp.getState())) {
+					SchedulingIntermediateDataSet intermediateDataSet = intermediateDataSets.get(srp.getResultId());
+					if (intermediateDataSet == null) {
+						throw new IllegalStateException("can not find scheduling intermediate dataset for "
+							+ srp.getResultId());
+					}
+					intermediateDataSet.incrementNumberOfRunningProducers();
+				}
+			}
+		}
+
+		List<ExecutionVertexDeploymentOption> executionVertexDeploymentOptions = new ArrayList<>(verticesToRestart.size());
+		for (ExecutionVertexID executionVertexId : verticesToRestart) {
+			final SchedulingExecutionVertex schedulingVertex = schedulingTopology.getVertex(executionVertexId)
+				.orElseThrow(() -> new IllegalStateException("can not find scheduling vertex for " + executionVertexId));
+			final InputDependencyConstraint constraint = schedulingTopology
+				.getInputDependencyConstraint(schedulingVertex.getId().getJobVertexId())
+				.orElseThrow(() -> new IllegalStateException("can not find input dependency constraint for "
+					+ executionVertexId));
+
+			Collection<SchedulingResultPartition> resultPartitions = schedulingVertex.getConsumedResultPartitions();
+			boolean allConsumable = resultPartitions.isEmpty() ||
+				(resultPartitions.iterator().next().getPartitionType().isBlocking() && InputDependencyConstraint.ALL.equals(constraint));
```

Comment:
Why is it required to check if the partition type is blocking, i.e., why is `InputDependencyConstraint.ALL.equals(constraint))` not enough? Why are we only checking the first result partition (`resultPartitions.iterator().next()`)?

Reply:
OK, `resultPartitions.isEmpty() || InputDependencyConstraint.ALL.equals(constraint)` is enough.